tests/apps/asgi/commands/test_new.py

```diff
@@ -1,28 +1,92 @@
+                    "piccolo.apps.asgi.commands.new.get_server",
+                    return_value=server,
+                ):
+                    root = os.path.join(tempfile.gettempdir(), "asgi_app")
 
-        if os.path.exists(root):
-            shutil.rmtree(root)
+                    if os.path.exists(root):
+                        shutil.rmtree(root)
 
-        os.mkdir(root)
-        new(root=root)
+                    os.mkdir(root)
+                    new(root=root)
 
-        self.assertTrue(os.path.exists(os.path.join(root, "app.py")))
+                    self.assertTrue(
+                        os.path.exists(os.path.join(root, "app.py"))
+                    )
+
+    def test_new_created_files_content(self, *args, **kwargs):
+        """
+        Test that the created files have the correct content.
+        list all .py files inside root directory and check
+         if they are valid python code with ast.parse
+        """
+        for router in ROUTERS:
+            for server in SERVERS:
+                with patch(
+                    "piccolo.apps.asgi.commands.new.get_routing_framework",
+                    return_value=router,
+                ), patch(
+                    "piccolo.apps.asgi.commands.new.get_server",
+                    return_value=server,
+                ):
+                    root = os.path.join(tempfile.gettempdir(), "asgi_app")
+
+                    if os.path.exists(root):
+                        shutil.rmtree(root)
+
+                    os.mkdir(root)
+                    new(root=root)
+
+                    for file in list(Path(root).rglob("*.py")):
+                        with open(os.path.join(root, file), "r") as f:
+                            ast.parse(f.read())
+                            f.close()
+
+    def test_run_new_app(self, *args, **kwargs):
+        """
+        test that the created app runs correctly.
+        """
+        for router in ROUTERS:
+            for server in SERVERS:
+                with patch(
+                    "piccolo.apps.asgi.commands.new.get_routing_framework",
+                    return_value=router,
+                ), patch(
+                    "piccolo.apps.asgi.commands.new.get_server",
+                    return_value=server,
+                ):
+                    root = os.path.join(tempfile.gettempdir(), "asgi_app")
+
+                    if os.path.exists(root):
+                        shutil.rmtree(root)
+
+                    os.mkdir(root)
+                    new(root=root)
+
+                    venv_path = os.path.join(root, "venv")
+                    os.mkdir(venv_path)
+                    os.system(
+                        f"#!/bin/bash &&"
+                        f"cd {root} && virtualenv venv &&"
+                        f" source {venv_path}/bin/activate &&"
+                        f" pip install -r requirements.txt &&"
+                        f" python main.py"
+                    )
```

Comment:
This looks great, thanks!

I wonder what happens here with `os.system`? Does the sub process just get killed off when the tests finish? I wonder if there's some way of getting a status code back from it?

Reply:
yes, the subprocess ends and I think it's possible to get the status code.
when everything is okay, it returns `0`. 
but try/except doesn't work with `os.system`.
using `subprocess` module gives more flexibility.
if you have anything in particular in your mind, I can try it out.

Reply:
@dantownsend for me looks good cause here if we use the try/except we could get multi issues when we debug, the idea that Ali provide using a subprocess is amazing.